File: typescript/design-patterns/builder/builder/RequestBuilder.ts
// Each step returns 'this' so calls chain.
class RequestBuilder {
  private req: HttpRequest = { method: "GET", url: "", headers: {} };

  method(m: string)            { this.req.method = m; return this; }
  url(u: string)               { this.req.url = u; return this; }
  header(k: string, v: string) { this.req.headers[k] = v; return this; }
  body(b: string)              { this.req.body = b; return this; }

  build(): HttpRequest {
    if (!this.req.url) throw new Error("url is required");
    return this.req;           // validate, then hand over
  }
}

File: java/design-patterns/builder/builder/RequestBuilder.java
public class RequestBuilder {
    private String method = "GET", url, body;
    private final Map<String,String> headers = new HashMap<>();

    public RequestBuilder method(String m) { this.method = m; return this; }
    public RequestBuilder url(String u)    { this.url = u;    return this; }
    public RequestBuilder header(String k, String v) {
        headers.put(k, v); return this;
    }
    public RequestBuilder body(String b)   { this.body = b;   return this; }

    public HttpRequest build() {
        if (url == null) throw new IllegalStateException("url required");
        return new HttpRequest(method, url, headers, body);
    }
}